// osmojs/esm/osmosis/smartaccount/v1beta1/tx.registry.js
import { MsgAddAuthenticator, MsgRemoveAuthenticator, MsgSetActiveState } from "./tx";
export const registry = [["/osmosis.smartaccount.v1beta1.MsgAddAuthenticator", MsgAddAuthenticator], ["/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator", MsgRemoveAuthenticator], ["/osmosis.smartaccount.v1beta1.MsgSetActiveState", MsgSetActiveState]];
export const load = (protoRegistry) => {
    registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
export const MessageComposer = {
    encoded: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value: MsgAddAuthenticator.encode(value).finish()
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value: MsgRemoveAuthenticator.encode(value).finish()
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value: MsgSetActiveState.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value
            };
        }
    },
    fromPartial: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value: MsgAddAuthenticator.fromPartial(value)
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value: MsgRemoveAuthenticator.fromPartial(value)
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value: MsgSetActiveState.fromPartial(value)
            };
        }
    }
};
