// osmojs/esm/ibc/applications/interchain_accounts/host/v1/tx.registry.js
import { MsgUpdateParams, MsgModuleQuerySafe } from "./tx";
export const registry = [["/ibc.applications.interchain_accounts.host.v1.MsgUpdateParams", MsgUpdateParams], ["/ibc.applications.interchain_accounts.host.v1.MsgModuleQuerySafe", MsgModuleQuerySafe]];
export const load = (protoRegistry) => {
    registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
export const MessageComposer = {
    encoded: {
        updateParams(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgUpdateParams",
                value: MsgUpdateParams.encode(value).finish()
            };
        },
        moduleQuerySafe(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgModuleQuerySafe",
                value: MsgModuleQuerySafe.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        updateParams(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgUpdateParams",
                value
            };
        },
        moduleQuerySafe(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgModuleQuerySafe",
                value
            };
        }
    },
    fromPartial: {
        updateParams(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgUpdateParams",
                value: MsgUpdateParams.fromPartial(value)
            };
        },
        moduleQuerySafe(value) {
            return {
                typeUrl: "/ibc.applications.interchain_accounts.host.v1.MsgModuleQuerySafe",
                value: MsgModuleQuerySafe.fromPartial(value)
            };
        }
    }
};
