// osmojs/osmosis/txfees/v1beta1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/osmosis.txfees.v1beta1.MsgSetFeeTokens", tx_1.MsgSetFeeTokens]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        setFeeTokens(value) {
            return {
                typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokens",
                value: tx_1.MsgSetFeeTokens.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        setFeeTokens(value) {
            return {
                typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokens",
                value
            };
        }
    },
    fromPartial: {
        setFeeTokens(value) {
            return {
                typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokens",
                value: tx_1.MsgSetFeeTokens.fromPartial(value)
            };
        }
    }
};
