// osmojs/ibc/lightclients/wasm/v1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/ibc.lightclients.wasm.v1.MsgStoreCode", tx_1.MsgStoreCode], ["/ibc.lightclients.wasm.v1.MsgRemoveChecksum", tx_1.MsgRemoveChecksum], ["/ibc.lightclients.wasm.v1.MsgMigrateContract", tx_1.MsgMigrateContract]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        storeCode(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgStoreCode",
                value: tx_1.MsgStoreCode.encode(value).finish()
            };
        },
        removeChecksum(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgRemoveChecksum",
                value: tx_1.MsgRemoveChecksum.encode(value).finish()
            };
        },
        migrateContract(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgMigrateContract",
                value: tx_1.MsgMigrateContract.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        storeCode(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgStoreCode",
                value
            };
        },
        removeChecksum(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgRemoveChecksum",
                value
            };
        },
        migrateContract(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgMigrateContract",
                value
            };
        }
    },
    fromPartial: {
        storeCode(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgStoreCode",
                value: tx_1.MsgStoreCode.fromPartial(value)
            };
        },
        removeChecksum(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgRemoveChecksum",
                value: tx_1.MsgRemoveChecksum.fromPartial(value)
            };
        },
        migrateContract(value) {
            return {
                typeUrl: "/ibc.lightclients.wasm.v1.MsgMigrateContract",
                value: tx_1.MsgMigrateContract.fromPartial(value)
            };
        }
    }
};
