// osmojs/osmosis/smartaccount/v1beta1/query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAuthenticatorsResponse = exports.GetAuthenticatorsRequest = exports.QueryParamsResponse = exports.QueryParamsRequest = void 0;
//@ts-nocheck
const params_1 = require("./params");
const models_1 = require("./models");
const binary_1 = require("../../../binary");
const registry_1 = require("../../../registry");
function createBaseQueryParamsRequest() {
    return {};
}
exports.QueryParamsRequest = {
    typeUrl: "/osmosis.smartaccount.v1beta1.QueryParamsRequest",
    aminoType: "osmosis/smartaccount/query-params-request",
    is(o) {
        return o && o.$typeUrl === exports.QueryParamsRequest.typeUrl;
    },
    isSDK(o) {
        return o && o.$typeUrl === exports.QueryParamsRequest.typeUrl;
    },
    isAmino(o) {
        return o && o.$typeUrl === exports.QueryParamsRequest.typeUrl;
    },
    encode(_, writer = binary_1.BinaryWriter.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryParamsRequest();
        return message;
    },
    fromAmino(_) {
        const message = createBaseQueryParamsRequest();
        return message;
    },
    toAmino(_) {
        const obj = {};
        return obj;
    },
    fromAminoMsg(object) {
        return exports.QueryParamsRequest.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/smartaccount/query-params-request",
            value: exports.QueryParamsRequest.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return exports.QueryParamsRequest.decode(message.value);
    },
    toProto(message) {
        return exports.QueryParamsRequest.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.smartaccount.v1beta1.QueryParamsRequest",
            value: exports.QueryParamsRequest.encode(message).finish()
        };
    }
};
registry_1.GlobalDecoderRegistry.register(exports.QueryParamsRequest.typeUrl, exports.QueryParamsRequest);
registry_1.GlobalDecoderRegistry.registerAminoProtoMapping(exports.QueryParamsRequest.aminoType, exports.QueryParamsRequest.typeUrl);
function createBaseQueryParamsResponse() {
    return {
        params: params_1.Params.fromPartial({})
    };
}
exports.QueryParamsResponse = {
    typeUrl: "/osmosis.smartaccount.v1beta1.QueryParamsResponse",
    aminoType: "osmosis/smartaccount/query-params-response",
    is(o) {
        return o && (o.$typeUrl === exports.QueryParamsResponse.typeUrl || params_1.Params.is(o.params));
    },
    isSDK(o) {
        return o && (o.$typeUrl === exports.QueryParamsResponse.typeUrl || params_1.Params.isSDK(o.params));
    },
    isAmino(o) {
        return o && (o.$typeUrl === exports.QueryParamsResponse.typeUrl || params_1.Params.isAmino(o.params));
    },
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.params !== undefined) {
            params_1.Params.encode(message.params, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.params = params_1.Params.decode(reader, reader.uint32());
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseQueryParamsResponse();
        message.params = object.params !== undefined && object.params !== null ? params_1.Params.fromPartial(object.params) : undefined;
        return message;
    },
    fromAmino(object) {
        const message = createBaseQueryParamsResponse();
        if (object.params !== undefined && object.params !== null) {
            message.params = params_1.Params.fromAmino(object.params);
        }
        return message;
    },
    toAmino(message) {
        const obj = {};
        obj.params = message.params ? params_1.Params.toAmino(message.params) : undefined;
        return obj;
    },
    fromAminoMsg(object) {
        return exports.QueryParamsResponse.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/smartaccount/query-params-response",
            value: exports.QueryParamsResponse.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return exports.QueryParamsResponse.decode(message.value);
    },
    toProto(message) {
        return exports.QueryParamsResponse.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.smartaccount.v1beta1.QueryParamsResponse",
            value: exports.QueryParamsResponse.encode(message).finish()
        };
    }
};
registry_1.GlobalDecoderRegistry.register(exports.QueryParamsResponse.typeUrl, exports.QueryParamsResponse);
registry_1.GlobalDecoderRegistry.registerAminoProtoMapping(exports.QueryParamsResponse.aminoType, exports.QueryParamsResponse.typeUrl);
function createBaseGetAuthenticatorsRequest() {
    return {
        account: ""
    };
}
exports.GetAuthenticatorsRequest = {
    typeUrl: "/osmosis.smartaccount.v1beta1.GetAuthenticatorsRequest",
    aminoType: "osmosis/smartaccount/get-authenticators-request",
    is(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsRequest.typeUrl || typeof o.account === "string");
    },
    isSDK(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsRequest.typeUrl || typeof o.account === "string");
    },
    isAmino(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsRequest.typeUrl || typeof o.account === "string");
    },
    encode(message, writer = binary_1.BinaryWriter.create()) {
        if (message.account !== "") {
            writer.uint32(10).string(message.account);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAuthenticatorsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.account = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseGetAuthenticatorsRequest();
        message.account = object.account ?? "";
        return message;
    },
    fromAmino(object) {
        const message = createBaseGetAuthenticatorsRequest();
        if (object.account !== undefined && object.account !== null) {
            message.account = object.account;
        }
        return message;
    },
    toAmino(message) {
        const obj = {};
        obj.account = message.account === "" ? undefined : message.account;
        return obj;
    },
    fromAminoMsg(object) {
        return exports.GetAuthenticatorsRequest.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/smartaccount/get-authenticators-request",
            value: exports.GetAuthenticatorsRequest.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return exports.GetAuthenticatorsRequest.decode(message.value);
    },
    toProto(message) {
        return exports.GetAuthenticatorsRequest.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.smartaccount.v1beta1.GetAuthenticatorsRequest",
            value: exports.GetAuthenticatorsRequest.encode(message).finish()
        };
    }
};
registry_1.GlobalDecoderRegistry.register(exports.GetAuthenticatorsRequest.typeUrl, exports.GetAuthenticatorsRequest);
registry_1.GlobalDecoderRegistry.registerAminoProtoMapping(exports.GetAuthenticatorsRequest.aminoType, exports.GetAuthenticatorsRequest.typeUrl);
function createBaseGetAuthenticatorsResponse() {
    return {
        accountAuthenticators: []
    };
}
exports.GetAuthenticatorsResponse = {
    typeUrl: "/osmosis.smartaccount.v1beta1.GetAuthenticatorsResponse",
    aminoType: "osmosis/smartaccount/get-authenticators-response",
    is(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsResponse.typeUrl || Array.isArray(o.accountAuthenticators) && (!o.accountAuthenticators.length || models_1.AccountAuthenticator.is(o.accountAuthenticators[0])));
    },
    isSDK(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsResponse.typeUrl || Array.isArray(o.account_authenticators) && (!o.account_authenticators.length || models_1.AccountAuthenticator.isSDK(o.account_authenticators[0])));
    },
    isAmino(o) {
        return o && (o.$typeUrl === exports.GetAuthenticatorsResponse.typeUrl || Array.isArray(o.account_authenticators) && (!o.account_authenticators.length || models_1.AccountAuthenticator.isAmino(o.account_authenticators[0])));
    },
    encode(message, writer = binary_1.BinaryWriter.create()) {
        for (const v of message.accountAuthenticators) {
            models_1.AccountAuthenticator.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof binary_1.BinaryReader ? input : new binary_1.BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAuthenticatorsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.accountAuthenticators.push(models_1.AccountAuthenticator.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseGetAuthenticatorsResponse();
        message.accountAuthenticators = object.accountAuthenticators?.map(e => models_1.AccountAuthenticator.fromPartial(e)) || [];
        return message;
    },
    fromAmino(object) {
        const message = createBaseGetAuthenticatorsResponse();
        message.accountAuthenticators = object.account_authenticators?.map(e => models_1.AccountAuthenticator.fromAmino(e)) || [];
        return message;
    },
    toAmino(message) {
        const obj = {};
        if (message.accountAuthenticators) {
            obj.account_authenticators = message.accountAuthenticators.map(e => e ? models_1.AccountAuthenticator.toAmino(e) : undefined);
        }
        else {
            obj.account_authenticators = message.accountAuthenticators;
        }
        return obj;
    },
    fromAminoMsg(object) {
        return exports.GetAuthenticatorsResponse.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/smartaccount/get-authenticators-response",
            value: exports.GetAuthenticatorsResponse.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return exports.GetAuthenticatorsResponse.decode(message.value);
    },
    toProto(message) {
        return exports.GetAuthenticatorsResponse.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.smartaccount.v1beta1.GetAuthenticatorsResponse",
            value: exports.GetAuthenticatorsResponse.encode(message).finish()
        };
    }
};
registry_1.GlobalDecoderRegistry.register(exports.GetAuthenticatorsResponse.typeUrl, exports.GetAuthenticatorsResponse);
registry_1.GlobalDecoderRegistry.registerAminoProtoMapping(exports.GetAuthenticatorsResponse.aminoType, exports.GetAuthenticatorsResponse.typeUrl);
