// osmojs/osmosis/smartaccount/v1beta1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/osmosis.smartaccount.v1beta1.MsgAddAuthenticator", tx_1.MsgAddAuthenticator], ["/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator", tx_1.MsgRemoveAuthenticator], ["/osmosis.smartaccount.v1beta1.MsgSetActiveState", tx_1.MsgSetActiveState]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value: tx_1.MsgAddAuthenticator.encode(value).finish()
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value: tx_1.MsgRemoveAuthenticator.encode(value).finish()
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value: tx_1.MsgSetActiveState.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value
            };
        }
    },
    fromPartial: {
        addAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgAddAuthenticator",
                value: tx_1.MsgAddAuthenticator.fromPartial(value)
            };
        },
        removeAuthenticator(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgRemoveAuthenticator",
                value: tx_1.MsgRemoveAuthenticator.fromPartial(value)
            };
        },
        setActiveState(value) {
            return {
                typeUrl: "/osmosis.smartaccount.v1beta1.MsgSetActiveState",
                value: tx_1.MsgSetActiveState.fromPartial(value)
            };
        }
    }
};
