// osmojs/esm/osmosis/txfees/v1beta1/tx.js
//@ts-nocheck
import { FeeToken } from "./feetoken";
import { BinaryReader, BinaryWriter } from "../../../binary";
import { GlobalDecoderRegistry } from "../../../registry";
function createBaseMsgSetFeeTokens() {
    return {
        feeTokens: [],
        sender: ""
    };
}
export const MsgSetFeeTokens = {
    typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokens",
    aminoType: "osmosis/set-fee-tokens",
    is(o) {
        return o && (o.$typeUrl === MsgSetFeeTokens.typeUrl || Array.isArray(o.feeTokens) && (!o.feeTokens.length || FeeToken.is(o.feeTokens[0])) && typeof o.sender === "string");
    },
    isSDK(o) {
        return o && (o.$typeUrl === MsgSetFeeTokens.typeUrl || Array.isArray(o.fee_tokens) && (!o.fee_tokens.length || FeeToken.isSDK(o.fee_tokens[0])) && typeof o.sender === "string");
    },
    isAmino(o) {
        return o && (o.$typeUrl === MsgSetFeeTokens.typeUrl || Array.isArray(o.fee_tokens) && (!o.fee_tokens.length || FeeToken.isAmino(o.fee_tokens[0])) && typeof o.sender === "string");
    },
    encode(message, writer = BinaryWriter.create()) {
        for (const v of message.feeTokens) {
            FeeToken.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.sender !== "") {
            writer.uint32(18).string(message.sender);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgSetFeeTokens();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.feeTokens.push(FeeToken.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.sender = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        const message = createBaseMsgSetFeeTokens();
        message.feeTokens = object.feeTokens?.map(e => FeeToken.fromPartial(e)) || [];
        message.sender = object.sender ?? "";
        return message;
    },
    fromAmino(object) {
        const message = createBaseMsgSetFeeTokens();
        message.feeTokens = object.fee_tokens?.map(e => FeeToken.fromAmino(e)) || [];
        if (object.sender !== undefined && object.sender !== null) {
            message.sender = object.sender;
        }
        return message;
    },
    toAmino(message) {
        const obj = {};
        if (message.feeTokens) {
            obj.fee_tokens = message.feeTokens.map(e => e ? FeeToken.toAmino(e) : undefined);
        }
        else {
            obj.fee_tokens = message.feeTokens;
        }
        obj.sender = message.sender === "" ? undefined : message.sender;
        return obj;
    },
    fromAminoMsg(object) {
        return MsgSetFeeTokens.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/set-fee-tokens",
            value: MsgSetFeeTokens.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return MsgSetFeeTokens.decode(message.value);
    },
    toProto(message) {
        return MsgSetFeeTokens.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokens",
            value: MsgSetFeeTokens.encode(message).finish()
        };
    }
};
GlobalDecoderRegistry.register(MsgSetFeeTokens.typeUrl, MsgSetFeeTokens);
GlobalDecoderRegistry.registerAminoProtoMapping(MsgSetFeeTokens.aminoType, MsgSetFeeTokens.typeUrl);
function createBaseMsgSetFeeTokensResponse() {
    return {};
}
export const MsgSetFeeTokensResponse = {
    typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokensResponse",
    aminoType: "osmosis/txfees/set-fee-tokens-response",
    is(o) {
        return o && o.$typeUrl === MsgSetFeeTokensResponse.typeUrl;
    },
    isSDK(o) {
        return o && o.$typeUrl === MsgSetFeeTokensResponse.typeUrl;
    },
    isAmino(o) {
        return o && o.$typeUrl === MsgSetFeeTokensResponse.typeUrl;
    },
    encode(_, writer = BinaryWriter.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMsgSetFeeTokensResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseMsgSetFeeTokensResponse();
        return message;
    },
    fromAmino(_) {
        const message = createBaseMsgSetFeeTokensResponse();
        return message;
    },
    toAmino(_) {
        const obj = {};
        return obj;
    },
    fromAminoMsg(object) {
        return MsgSetFeeTokensResponse.fromAmino(object.value);
    },
    toAminoMsg(message) {
        return {
            type: "osmosis/txfees/set-fee-tokens-response",
            value: MsgSetFeeTokensResponse.toAmino(message)
        };
    },
    fromProtoMsg(message) {
        return MsgSetFeeTokensResponse.decode(message.value);
    },
    toProto(message) {
        return MsgSetFeeTokensResponse.encode(message).finish();
    },
    toProtoMsg(message) {
        return {
            typeUrl: "/osmosis.txfees.v1beta1.MsgSetFeeTokensResponse",
            value: MsgSetFeeTokensResponse.encode(message).finish()
        };
    }
};
GlobalDecoderRegistry.register(MsgSetFeeTokensResponse.typeUrl, MsgSetFeeTokensResponse);
GlobalDecoderRegistry.registerAminoProtoMapping(MsgSetFeeTokensResponse.aminoType, MsgSetFeeTokensResponse.typeUrl);
